bpf: lxc: ICMP fragmentation fix

m0r1k opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

This patch fixes the problem with sending a large ICMP packet with a payload up to 65507 bytes (the packet will be fragmented). As you know, ipv4 fragments do not contain L4 header (but we try to read it and as a result we read garbage from icmp payload)

Without ICMP data "type", "code", "identifier" we will not be able to do reverse nat for returning IP/ICMP fragments.

So I added a function that saves "type", "code", "identifier" values in IPV4_FRAG_DATAGRAMS_MAP and restores them for returning IP/ICMP fragments. So revert nat is executed successfully and echo reply returns inside the pod.

Information about outgoing ICMP fragments will be stored in ipv4_frag_l4ports as follows:

```
ports.sport = (__be16)((type << 8) | code);
ports.dport = identifier;
```

before:
![image](https://github.com/user-attachments/assets/75b0872d-500c-4623-a0ae-be6f729077b6)

after:
![image](https://github.com/user-attachments/assets/7e10b461-1063-4364-8562-282908812df7)

lab:
![image](https://github.com/user-attachments/assets/73c4b555-e2a6-4870-bd6a-baeb02e62ff8)

test (egress echo):
![image](https://github.com/user-attachments/assets/563bc3fb-d2db-4833-9989-3afce1d355f7)

test (echo reply):
![image](https://github.com/user-attachments/assets/65c4ec12-d04e-4471-92ca-8a3c60e7be70)

```release-note
fixed the problem with sending a large ICMP packet with a payload up to 65507 bytes (icmp fragmentation)
```

Commit:
bpf: lxc: ICMP fragmentation fix

This patch fixes the problem with sending a large ICMP packet
with a payload up to 65507 bytes (the packet will be fragmented).
As you know, ipv4 fragments do not contain L4 header
(but we try to read it and as a result we read garbage from icmp payload)

Without ICMP data "type", "code", "identifier" we will not be able
to do reverse nat for returning IP/ICMP fragments.

So I added a function that saves "type", "code", "identifier" values
in IPV4_FRAG_DATAGRAMS_MAP and restores them for returning IP/ICMP fragments.
So revert nat is executed successfully and echo reply returns inside the pod.

Information about outgoing ICMP fragments will be stored
in ipv4_frag_l4ports as follows:

ports.sport = (__be16)((type << 8) | code);
ports.dport = identifier;

Signed-off-by: Roman E. Chechnev <roman.chechnev@flant.ru>

gentoo-root commented on bpf/lib/ipv4.h:
```diff
@@ -174,10 +210,132 @@ ipv4_handle_fragmentation(struct __ctx_buff *ctx,
+	struct __ctx_buff	*a_ctx,
+	int					a_l4_off,
+	enum ct_dir			a_ct_dir __maybe_unused,
+	__u8				*a_type,
```
We also don't seem to add the `a_` prefix to all function parameters, so please follow the existing style. While it can be meaningful in some other project, where it's explicitly explained in the coding style guide, here it looks out of context.

m0r1k replied:
how to understand inside the function, which variable is argument, which variable is property, and which variable is local, for this, prefixes are used a_ (argument), s_ (static), m_ (member) g_ (global) but it does not matter to you, because in your "codestyle" is allowed even rewrite incoming arguments ))

![image](https://github.com/user-attachments/assets/2209b9d1-95b7-4179-8a8c-5b26bebdc5bf)

gentoo-root commented on bpf/lib/ipv4.h:
```diff
@@ -174,10 +210,132 @@ ipv4_handle_fragmentation(struct __ctx_buff *ctx,
+	__be16				*a_identifier,
+	bool				a_create_frag_record)
+{
+	int							ret			= 0;
```
Please address all [checkpatch errors](https://github.com/cilium/cilium/actions/runs/13010942666/job/36288366930?pr=37306) and follow the existing code style in our project. We don't do "table" alignment of variable declarations and function parameters in C code (especially when the result doesn't even look like a table :slightly_smiling_face:). When unsure, refer to [Linux coding style](https://www.kernel.org/doc/html/latest/process/coding-style.html) or look at the code around.

m0r1k replied:
> We don't do "table" alignment of variable declarations 

you even can't do it, because your "codestyle" is the mix of spaces and tabs ))

spaces:
![image](https://github.com/user-attachments/assets/7be58ef1-6b23-44f8-857f-af8278be9b5a)

tabs:
![image](https://github.com/user-attachments/assets/0016e634-8264-46d8-8729-a4f21ad7efba)

> When unsure, refer to [Linux coding style](https://www.kernel.org/doc/html/latest/process/coding-style.html) or look at the code around.

I refer to my mind and KISS rule, one line - one thought and I have not seen worse code for 20 years in Linux

gentoo-root commented on bpf/lib/ipv4.h:
```diff
@@ -174,10 +210,132 @@ ipv4_handle_fragmentation(struct __ctx_buff *ctx,
+	}
+#endif
+
+	*a_type			= type;
```
You can write directly to `*a_type`, there is no need for the intermediate local variable, as the function is marked as `__always_inline`, and there won't be an actual pointer dereference.

m0r1k replied:
the code must be readable

gentoo-root reviewed (changes requested):
Thanks for your contribution! Please clean up the coding style according to the one used in our project, before I do a meaningful review.

Commit:
bpf: lxc: ICMP fragmentation fix

This patch fixes the problem with sending a large ICMP packet
with a payload up to 65507 bytes (the packet will be fragmented).
As you know, ipv4 fragments do not contain L4 header
(but we try to read it and as a result we read garbage from icmp payload)

Without ICMP data "type", "code", "identifier" we will not be able
to do reverse nat for returning IP/ICMP fragments.

So I added a function that saves "type", "code", "identifier" values
in IPV4_FRAG_DATAGRAMS_MAP and restores them for returning IP/ICMP fragments.
So revert nat is executed successfully and echo reply returns inside the pod.

Information about outgoing ICMP fragments will be stored
in ipv4_frag_l4ports as follows:

ports.sport = (__be16)((type << 8) | code);
ports.dport = identifier;

Signed-off-by: Roman E. Chechnev <roman.chechnev@flant.ru>

Commit:
bpf: lxc: ICMP fragmentation fix

This patch fixes the problem with sending a large ICMP packet
with a payload up to 65507 bytes (the packet will be fragmented).
As you know, ipv4 fragments do not contain L4 header
(but we try to read it and as a result we read garbage from icmp payload)

Without ICMP data "type", "code", "identifier" we will not be able
to do reverse nat for returning IP/ICMP fragments.

So I added a function that saves "type", "code", "identifier" values
in IPV4_FRAG_DATAGRAMS_MAP and restores them for returning IP/ICMP fragments.
So revert nat is executed successfully and echo reply returns inside the pod.

Information about outgoing ICMP fragments will be stored
in ipv4_frag_l4ports as follows:

ports.sport = (__be16)((type << 8) | code);
ports.dport = identifier;

Signed-off-by: Roman E. Chechnev <roman.chechnev@flant.ru>

squeed commented on bpf/lib/policy.h:
```diff
@@ -82,21 +82,29 @@ __policy_can_access(const void *map, struct __ctx_buff *ctx, __u32 local_id,
+				&type, &code, &identifier,
+				false // create_frag_record
+			);
+			if (ret < 0)
```
This change newly adds the ability for policy calculation to fail; previously this was not the case. I'm not sure all the callers really handle that properly.

For the TCP / UDP case, we look up the numbers from conntrack in `ipv4_policy()`. Let's do the same for ICMP.

gentoo-root replied:
I don't see Casey's suggestion implemented.

If we are keeping the existing logic, though, it can be refactored to use icmp4_load_info_from_header.

Also, if we are looking for ICMP_DEST_UNREACH only, we shouldn't need to look up the fragments, as long as the [statement in this comment](https://github.com/cilium/cilium/pull/37306#discussion_r1937065813) holds true.

gentoo-root commented on bpf/lib/conntrack.h:
```diff
@@ -913,7 +913,7 @@ static __always_inline int ct_lookup4(const void *map,
 {
 	bool is_fragment = ipv4_is_fragment(ip4);
-	bool has_l4_header = true;
+	bool has_l4_header = ipv4_has_l4_header(ip4);
```
Is this change needed? has_l4_header is an output parameter of ct_extract_ports4 and ipv4_handle_fragmentation. When ENABLE_IPV4_FRAGMENTS is set, has_l4_header will be set upon return from ct_extract_ports. When ENABLE_IPV4_FRAGMENTS is unset, fragmented packets are dropped before entering ct_lookup4.

gentoo-root commented on bpf/lib/ipv4.h:
```diff
@@ -174,10 +209,120 @@ ipv4_handle_fragmentation(struct __ctx_buff *ctx,
+#endif
+
+	if (NULL == type_out ||	NULL == code_out || NULL == identifier_out) {
+		ret = EINVAL;
```
Error codes are negative, it should be `-EINVAL`. We also return specific drop reasons (DROP_* codes) from functions that go to send_drop_notify_error and then can be deciphered by tools such as cilium monitor. We don't propagate raw kernel error codes there. In cases where we need them (to distinguish underlying errors), we use ext_err.

gentoo-root commented on bpf/lib/ipv4.h:
```diff
@@ -174,10 +209,120 @@ ipv4_handle_fragmentation(struct __ctx_buff *ctx,
+	enum metric_dir mdir = ct_to_metrics_dir(ct_dir);
+#endif
+
+	if (NULL == type_out ||	NULL == code_out || NULL == identifier_out) {
```
No need to NULL-check these, it just hides bugs. The verifier won't pass invalid code, such as when the caller passes NULL, and this function tries to deref it. With the NULL-check, however, the verifier would pass, and the bug would only be seen in runtime.

Wherever NULL checks are needed, they are done with `if (!something)`, not `if (NULL == something)`, checkpatch throws an error on this.

gentoo-root commented on bpf/lib/ipv4.h:
```diff
@@ -174,10 +209,120 @@ ipv4_handle_fragmentation(struct __ctx_buff *ctx,
-	return 0;
+out:
+	return ret;
+fail:
```
Two labels point at the same place, both are used on errors, and both don't do any additional cleanup. Please remove both labels and just return in place.

gentoo-root commented on bpf/lib/nat.h:
```diff
@@ -851,25 +850,31 @@ snat_v4_nat(struct __ctx_buff *ctx, struct ipv4_ct_tuple *tuple,
 			break;
 		case ICMP_ECHOREPLY:
 			return NAT_PUNT_TO_STACK;
 		case ICMP_DEST_UNREACH:
```
These types of ICMP packets (ICMP_DEST_UNREACH, ICMP_TIME_EXCEEDED) shouldn't be fragmented at the first place. Do you face a real-world situation where it happens? Or is it all about large pings?

gentoo-root commented on bpf/lib/ipv4.h:
```diff
@@ -174,10 +209,120 @@ ipv4_handle_fragmentation(struct __ctx_buff *ctx,
+	*code_out = code;
+	*identifier_out	= identifier;
+
+out:
```
Same here, just return on errors.

gentoo-root commented on bpf/lib/ipv4.h:
```diff
@@ -174,10 +209,120 @@ ipv4_handle_fragmentation(struct __ctx_buff *ctx,
+	}
+#endif
+
+	*type_out = type;
```
No need for an extra copy from intermediate variables, just write directly to `*type_out`, etc. (and drop `_out` from the name). The function is inline.

gentoo-root commented on bpf/lib/nat.h:
```diff
@@ -991,6 +996,8 @@ snat_v4_rev_nat(struct __ctx_buff *ctx, const struct ipv4_nat_target *target,
 
 	snat_v4_init_tuple(ip4, NAT_DIR_INGRESS, &tuple);
 
+	has_l4_header = ipv4_has_l4_header(ip4);
```
Same comment as the other occurrence of has_l4_header.

gentoo-root commented on bpf/lib/ipv4.h:
```diff
@@ -174,10 +209,120 @@ ipv4_handle_fragmentation(struct __ctx_buff *ctx,
+		goto out;
+	}
+
+	if (!revalidate_data(ctx, &data, &data_end, &ip4)) {
```
Avoid revalidate_data, pass ip4 externally. Aside from adding extra branches, it erases the verifier's knowledge of checks already done above, often requiring repeating those checks again.

gentoo-root commented on bpf/lib/ipv4.h:
```diff
@@ -174,10 +209,120 @@ ipv4_handle_fragmentation(struct __ctx_buff *ctx,
 #endif
 
+static __always_inline int
+ipv4_load_l4_ports_for_icmp(struct __ctx_buff *ctx, int l4_off, enum ct_dir ct_dir __maybe_unused,
```
I'd think of a better name for this function, type and code aren't ports in ICMP.

gentoo-root commented on bpf/lib/ipv4.h:
```diff
@@ -174,10 +209,120 @@ ipv4_handle_fragmentation(struct __ctx_buff *ctx,
+#ifdef ENABLE_IPV4_FRAGMENTS
+	if (!has_l4_header) {
+		/* it should be a fragmented packet */
+		is_fragment = ipv4_is_not_first_fragment(ip4);
```
`ipv4_has_l4_header` returns `!ipv4_is_not_first_fragment`, no need to check the same thing twice.

gentoo-root commented on bpf/lib/ipv4.h:
```diff
@@ -174,10 +209,120 @@ ipv4_handle_fragmentation(struct __ctx_buff *ctx,
+#endif
+		}
+	}
+#ifdef ENABLE_IPV4_FRAGMENTS
```
For better readability, please structure this function similarly to ipv4_load_l4_ports/ipv4_handle_fragmentation/ipv4_frag_get_l4ports/l4_load_ports. Smaller functions with well-defined purpose are better for readability (similar structure also helps), and `#ifdef`ing entire functions is both more readable than `#ifdef`ing every few lines, and better aligns with our clang-free plans.

gentoo-root commented on bpf/lib/ipv4.h:
```diff
@@ -174,10 +209,120 @@ ipv4_handle_fragmentation(struct __ctx_buff *ctx,
+
+	if (has_l4_header) {
+		/* load identifier from ICMP header */
+		if (ctx_load_bytes(ctx, l4_off, &type, 1) < 0) {
```
Duplicating code with `ipv4_handle_fragmentation()`. Please move it to a dedicated function. In TCP/UDP cases, l4_load_ports serves this purpose, see the comment below.

gentoo-root commented on bpf/lib/ipv4.h:
```diff
@@ -152,14 +153,48 @@ ipv4_handle_fragmentation(struct __ctx_buff *ctx,
-	if (ret < 0)
-		return DROP_CT_INVALID_HDR;
+	switch (ip4->protocol) {
+		case IPPROTO_ICMP: {
```
Is this entire branch actually useful here? It seems you go to `ipv4_load_l4_ports_for_icmp()` in ICMP cases, which doesn't call `ipv4_handle_fragmentation()`, unlike `ipv4_load_l4_ports()` does?

gentoo-root replied:
Same question again. ipv4_load_l4_ports is never called for ICMP packets (or did I miss something?), so ipv4_handle_fragmentation can't be called for ICMP.

gentoo-root commented on bpf/lib/ipv4.h:
```diff
@@ -174,10 +209,120 @@ ipv4_handle_fragmentation(struct __ctx_buff *ctx,
+			if (map_update_elem(&IPV4_FRAG_DATAGRAMS_MAP, &frag_id, &ports, BPF_ANY))
+				update_metrics(ctx_full_len(ctx), mdir, REASON_FRAG_PACKET_UPDATE);
+
+			/* Do not return an error if map update failed, as nothing prevents us */
```
The right formatting for multiline comments is: https://github.com/cilium/cilium/blob/46a49b899cccb352fec09a58b68ba8f709d8cf78/bpf/lib/ipv4.h#L172-L174

You copied this comment from existing code and formatted it differently on purpose. Following our coding style is a requirement, and non-conforming code simply can't be merged. Please make sure to clean up all remaining things.

Commit:
bpf: lxc: ICMP fragmentation fix

This patch fixes the problem with sending a large ICMP packet
with a payload up to 65507 bytes (the packet will be fragmented).
As you know, ipv4 fragments do not contain L4 header
(but we try to read it and as a result we read garbage from icmp payload)

Without ICMP data "type", "code", "identifier" we will not be able
to do reverse nat for returning IP/ICMP fragments.

So I added a function that saves "type", "code", "identifier" values
in IPV4_FRAG_DATAGRAMS_MAP and restores them for returning IP/ICMP fragments.
So revert nat is executed successfully and echo reply returns inside the pod.

Information about outgoing ICMP fragments will be stored
in ipv4_frag_l4ports as follows:

ports.sport = (__be16)((type << 8) | code);
ports.dport = identifier;

Signed-off-by: Roman E. Chechnev <roman.chechnev@flant.ru>

gentoo-root commented on bpf/lib/ipv4.h:
```diff
@@ -111,6 +112,33 @@ static __always_inline bool ipv4_is_in_subnet(__be32 addr,
+	/* load identifier from ICMP header */
+	if (ctx_load_bytes(ctx, l4_off, type_out, 1) < 0) {
+		ret = DROP_CT_INVALID_HDR;
+		goto out;
```
Return in place, no need for a label to a single return statement, also no need even for the ret variable. (Applies to all occurrences.)

gentoo-root commented on bpf/lib/ipv4.h:
```diff
@@ -111,6 +112,33 @@ static __always_inline bool ipv4_is_in_subnet(__be32 addr,
 
+static __always_inline int
+icmp4_load_info_from_header(struct __ctx_buff *ctx, const struct iphdr *ip4 __maybe_unused,
+			    int l4_off, __u8 *type_out, __u8 *code_out, __be16 *identifier_out)
```
Just *type, *code, *identifier please, no need for this suffix. Same for icmp4_handle_fragmentation.

gentoo-root commented on bpf/lib/ipv4.h:
```diff
@@ -111,6 +112,33 @@ static __always_inline bool ipv4_is_in_subnet(__be32 addr,
 }
 
+static __always_inline int
+icmp4_load_info_from_header(struct __ctx_buff *ctx, const struct iphdr *ip4 __maybe_unused,
```
ip4 is always unused, please remove.

gentoo-root commented on bpf/lib/ipv4.h:
```diff
@@ -174,7 +221,68 @@ ipv4_handle_fragmentation(struct __ctx_buff *ctx,
+		goto out;
+
+	if (is_fragment) {
+		/* "more fragments" flag is set, or we have frag offset */
```
[Same](https://github.com/cilium/cilium/pull/37306#discussion_r1937124293) comment for formatting of comments.

gentoo-root commented on bpf/lib/ipv4.h:
```diff
@@ -156,10 +184,29 @@ ipv4_handle_fragmentation(struct __ctx_buff *ctx,
+			ports->dport = identifier;
+			break;
+		}
+		default: {
```
switch labels should be indented at the same level as the switch itself.

gentoo-root reviewed (changes requested):
I also unresolved a bunch of old comments, because they haven't been addressed, and the questions haven't been answered. Please address accordingly.

Commit:
bpf: lxc: ICMP fragmentation fix

This patch fixes the problem with sending a large ICMP packet
with a payload up to 65507 bytes (the packet will be fragmented).
As you know, ipv4 fragments do not contain L4 header
(but we try to read it and as a result we read garbage from icmp payload)

Without ICMP data "type", "code", "identifier" we will not be able
to do reverse nat for returning IP/ICMP fragments.

So I added a function that saves "type", "code", "identifier" values
in IPV4_FRAG_DATAGRAMS_MAP and restores them for returning IP/ICMP fragments.
So revert nat is executed successfully and echo reply returns inside the pod.

Information about outgoing ICMP fragments will be stored
in ipv4_frag_l4ports as follows:

ports.sport = (__be16)((type << 8) | code);
ports.dport = identifier;

Signed-off-by: Roman E. Chechnev <roman.chechnev@flant.ru>

Commit:
bpf: lxc: ICMP fragmentation fix

This patch fixes the problem with sending a large ICMP packet
with a payload up to 65507 bytes (the packet will be fragmented).
As you know, ipv4 fragments do not contain L4 header
(but we try to read it and as a result we read garbage from icmp payload)

Without ICMP data "type", "code", "identifier" we will not be able
to do reverse nat for returning IP/ICMP fragments.

So I added a function that saves "type", "code", "identifier" values
in IPV4_FRAG_DATAGRAMS_MAP and restores them for returning IP/ICMP fragments.
So revert nat is executed successfully and echo reply returns inside the pod.

Information about outgoing ICMP fragments will be stored
in ipv4_frag_l4ports as follows:

ports.sport = (__be16)((type << 8) | code);
ports.dport = identifier;

Signed-off-by: Roman E. Chechnev <roman.chechnev@flant.ru>

Commit:
bpf: lxc: ICMP fragmentation fix

This patch fixes the problem with sending a large ICMP packet
with a payload up to 65507 bytes (the packet will be fragmented).
As you know, ipv4 fragments do not contain L4 header
(but we try to read it and as a result we read garbage from icmp payload)

Without ICMP data "type", "code", "identifier" we will not be able
to do reverse nat for returning IP/ICMP fragments.

So I added a function that saves "type", "code", "identifier" values
in IPV4_FRAG_DATAGRAMS_MAP and restores them for returning IP/ICMP fragments.
So revert nat is executed successfully and echo reply returns inside the pod.

Information about outgoing ICMP fragments will be stored
in ipv4_frag_l4ports as follows:

ports.sport = (__be16)((type << 8) | code);
ports.dport = identifier;

Signed-off-by: Roman E. Chechnev <roman.chechnev@flant.ru>
